core/commands/refs.go

```diff
@@ -74,65 +77,37 @@ NOTE: List all references recursively by using the flag '-r'.
 		cmdkit.BoolOption(refsRecursiveOptionName, "r", "Recursively list links of child nodes."),
 		cmdkit.IntOption(refsMaxDepthOptionName, "Only for recursive refs, limits fetch and listing to the given depth").WithDefault(-1),
 	},
-	Run: func(req cmds.Request, res cmds.Response) {
-		ctx := req.Context()
-		n, err := req.InvocContext().GetNode()
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+		ctx := req.Context
+		n, err := cmdenv.GetNode(env)
 		if err != nil {
-			res.SetError(err, cmdkit.ErrNormal)
-			return
+			return err
 		}
 
-		unique, _, err := req.Option(refsUniqueOptionName).Bool()
-		if err != nil {
-			res.SetError(err, cmdkit.ErrNormal)
-			return
-		}
-
-		recursive, _, err := req.Option(refsRecursiveOptionName).Bool()
-		if err != nil {
-			res.SetError(err, cmdkit.ErrNormal)
-			return
-		}
-
-		maxDepth, _, err := req.Option(refsMaxDepthOptionName).Int()
-		if err != nil {
-			res.SetError(err, cmdkit.ErrNormal)
-			return
-		}
+		unique, _ := req.Options[refsUniqueOptionName].(bool)
+		recursive, _ := req.Options[refsRecursiveOptionName].(bool)
+		maxDepth, _ := req.Options[refsMaxDepthOptionName].(int)
+		edges, _ := req.Options[refsEdgesOptionName].(bool)
+		format, _ := req.Options[refsFormatOptionName].(string)
 
 		if !recursive {
 			maxDepth = 1 // write only direct refs
 		}
 
-		format, _, err := req.Option(refsFormatOptionName).String()
-		if err != nil {
-			res.SetError(err, cmdkit.ErrNormal)
-			return
-		}
-
-		edges, _, err := req.Option(refsEdgesOptionName).Bool()
-		if err != nil {
-			res.SetError(err, cmdkit.ErrNormal)
-			return
-		}
 		if edges {
 			if format != "<dst>" {
-				res.SetError(errors.New("using format argument with edges is not allowed"),
-					cmdkit.ErrClient)
-				return
+				return errors.New("using format argument with edges is not allowed")
 			}
 
 			format = "<src> -> <dst>"
 		}
 
-		objs, err := objectsForPaths(ctx, n, req.Arguments())
+		objs, err := objectsForPaths(ctx, n, req.Arguments)
 		if err != nil {
-			res.SetError(err, cmdkit.ErrNormal)
-			return
+			return err
 		}
 
 		out := make(chan interface{})
```

Comment:
We should likely drop this channel and go routine and use res.Emit directly for each value?